Author: Mort65, https://github.com/mort65/
Description: This patch toggles the visibility of all windows on current monitor
diff --git a/config.def.h b/config.def.h
index ba9a240..eaaca3e 100644
--- a/config.def.h
+++ b/config.def.h
@@ -58,11 +58,14 @@ static const Layout layouts[] = {
 static char dmenumon[2] = "0"; /* component of dmenucmd, manipulated in spawn() */
 static const char *dmenucmd[] = { "dmenu_run", "-m", dmenumon, "-fn", dmenufont, "-nb", col_gray1, "-nf", col_gray3, "-sb", col_cyan, "-sf", col_gray4, NULL };
 static const char *termcmd[]  = { "st", NULL };

 static Key keys[] = {
 	/* modifier                     key        function        argument */
 	{ MODKEY,                       XK_p,      spawn,          {.v = dmenucmd } },
 	{ MODKEY|ShiftMask,             XK_Return, spawn,          {.v = termcmd } },
+	{ MODKEY,                       XK_z,      togglewindows,  {0} },
 	{ MODKEY,                       XK_b,      togglebar,      {0} },
 	{ MODKEY,                       XK_j,      focusstack,     {.i = +1 } },
 	{ MODKEY,                       XK_k,      focusstack,     {.i = -1 } },
diff --git a/dwm.c b/dwm.c
index d27cb67..3fa7274 100644
--- a/dwm.c
+++ b/dwm.c
@@ -212,6 +212,7 @@ static void tagmon(const Arg *arg);
 static void tile(Monitor *);
 static void togglebar(const Arg *arg);
 static void togglefloating(const Arg *arg);
+static void togglewindows(const Arg *arg);
 static void toggletag(const Arg *arg);
 static void toggleview(const Arg *arg);
 static void unfocus(Client *c, int setfocus);
@@ -272,6 +273,8 @@ static Window root, wmcheckwin;
 /* configuration, allows nested code to access above variables */
 #include "config.h"

+static unsigned int emptytag = 1 << (LENGTH(tags) + 1);
+
 /* compile-time check if all tags fit into an unsigned int bit array. */
 struct NumTags { char limitexceeded[LENGTH(tags) > 31 ? -1 : 1]; };

@@ -1051,6 +1054,13 @@ manage(Window w, XWindowAttributes *wa)
 	           && (c->x + (c->w / 2) < c->mon->wx + c->mon->ww)) ? bh : c->mon->my);
 	c->bw = borderpx;

+	if (c->mon->tagset[c->mon->seltags] >= emptytag)
+		c->tags = (c->tags % emptytag) | (c->tags >> (LENGTH(tags) + 1));
+
	wc.border_width = c->bw;
 	XConfigureWindow(dpy, w, CWBorderWidth, &wc);
 	XSetWindowBorder(dpy, w, scheme[SchemeNorm][ColBorder].pixel);
@@ -1724,6 +1734,28 @@ togglefloating(const Arg *arg)
 }

void
+togglewindows(const Arg *arg)
+{
+	if (selmon->tagset[selmon->seltags]) {
+		unsigned int newtagset = selmon->tagset[selmon->seltags] >= emptytag ? (selmon->tagset[selmon->seltags] % emptytag) | (selmon->tagset[selmon->seltags] >> (LENGTH(tags) + 1)) : selmon->tagset[selmon->seltags] << (LENGTH(tags) + 1);
+		selmon->tagset[selmon->seltags] = newtagset;
+		focus(NULL);
+		arrange(selmon);
+	}
+}
+
+void
 toggletag(const Arg *arg)
 {
 	unsigned int newtags;
